### DISCUSSION TITLE
about MagickWand ImageToBlob

### DISCUSSION THREAD

**Discussion by hellomsg** [ImageMagick/ImageMagick#2456] [Help] [answered]
I don't understand why the `MagickGetImageBlob` and `ImageToBlob` method has a `length` parameter.
I think the `length` should retrieval from the current image's image_info. The `MagickGetImageBlob` method has the ability do it self? 

And I don't know much about the MagickWand's design principles. A Wand has many images? But `MagickResizeImage` process which image in the wand image list? Where can I get these introduction?

Below is my tour code use swift.
newbie 😭 

```swift
#if canImport(Darwin)
import Darwin
#else
import Glibc
#endif
import Foundation
import CImageMagick

print("Hello, world!")
// magick wand
MagickWandGenesis()
let magick_wand = NewMagickWand()
let magick_wand2 = NewMagickWand()
var status: MagickBooleanType = MagickReadImage(magick_wand, "/Users/x/git/ImageMagickSwiftFoo/logo.gif")

if status == MagickFalse {
    print("load image error")
} else {
    MagickResetIterator(magick_wand)
    while MagickNextImage(magick_wand) != MagickFalse {
        let image = GetImageFromMagickWand(magick_wand)[0];
        MagickResizeImage(magick_wand, 106, 80, LanczosFilter)
        let blob = MagickGetImageBlob(magick_wand, &image.image_info[0].length);
        
        status = MagickReadImageBlob(magick_wand2, blob, image.image_info[0].length)
        print(status)
        status = MagickWriteImages(magick_wand, "/Users/x/git/ImageMagickSwiftFoo/resize.gif", MagickTrue)
        status = MagickWriteImages(magick_wand2, "/Users/x/git/ImageMagickSwiftFoo/resize2.gif", MagickTrue)
        if status == MagickFalse {
            print("write image error")
        }

    }
}
DestroyMagickWand(magick_wand2)
DestroyMagickWand(magick_wand)
MagickWandTerminus()
```

**Comment by snibgo** [comment, marked as answer]
The functions are documented at https://www.imagemagick.org/api/magick-image.php

MagickGetImageBlob() creates a blob, and returns the blob length into the location you give. Writing this length to a image_info element seems unwise. Writing to your own variable seems more sensible. MagickReadImageBlob() reads a blob, but you must tell it what length the blob is.

A wand has a list of images, and one of these is the current image. MagickResizeImage() etc operate on the current image.

**Reply by hellomsg** [reply]
oh yeah, length is a in-out parameter. Thank you 🥰
